### ISSUE TITLE
TX Values

### ISSUE THREAD

**Issue by horiajurcut** [google/exposure-notifications-internals#2]
Maybe this is not the write forum to ask this question but I've seen a lot of resources explaining the way Attenuation is calculated in Exposure Notification and none of them are very clear.

Could someone please explain where these two values come from? As I can see in the code they're being summed up and sent in the encrypted payload. Is txCalibrationPower the value from the calibration dataset? Then what exactly is the offset?

`/**
     * The transmit power as would be reported by an average iPhone
     * transmitting at the same power as the Android device running with
     * ADVERTISE_TX_POWER_LOW
     */
    public static int txCalibrationPower() {
        return 0;
    }

    /**
     * Offset to add to the tx calibration power to obtain the tx
     * power to publish in the payload
     */
    public static int txCalibrationPowerToTxPower() {
        return 0;
    }
`

According to the docs: Attenuation = TX_POWER - (RSSI_MEASURED + RSSI_CORRECTION)

Browsing this repo, it seems that TX_POWER is the sum of the above values (which I don't fully understand what they are) that are sent from transmitter to receiver via the encrypted payload.

RSSI_MEASURED = is the signal strength value measured by the receiver

RSSI_CORRECTION = is the value from the calibration dataset (do you take this value for the receiver or transmitter's oem and model?)

Any clarification would be very much appreciated! Thanks!

**Comment by hvisser** [comment]
Details on the EN framework BLE calibration and values for it are published here: https://developers.google.com/android/exposure-notifications/ble-attenuation-overview

My assumption is that within Google Play Services the actual calibration values that are listed there are used as config and that the methods you reference here use a stubbed out version here, hence `0`. But would be great to get some confirmation on that.

**Comment by horiajurcut** [comment]
@hvisser Yeah, that's what I've been reading, but it's still not clear what those values are. `txCalibrationPower()` seems to be a stub for the value in the calibration [CSV](https://developers.google.com/android/exposure-notifications/files/en-calibration-2020-06-13.csv) so what is `txCalibrationPowerToTxPower()`?

**Comment by hvisser** [comment]
Well, it's not the Bluetooth protocol one, since that isn't used, but a tx power value is broadcasted in the associated (encrypted) meta data of the advertisement. Rereading the comments `txCalibrationPowerToTxPower()` could be that value as broadcasted from the device, to tell other devices what the perceived tx power is, similar to the Bluetooth protocol one.

**Comment by horiajurcut** [comment]
It would be great if someone familiar with the internals could clarify what all these values represent and where are they read from. Thank you!

**Comment by horiajurcut** [comment]
Another thing I have found, a swiss experiment measuring the values of the transmission power in the encrypted payload for a Pixel 2. The value measured was -17. In the CSV calibration table the tx value for Pixel 2 is -32. The tx power advertised at POWER_LOW is -15. Adding these up like in the above code does not give us the desired outcome so I'm extremely curious now what these values are.

**Comment by mh-** [comment]
@horiajurcut 
> The value [...] was -17. In the CSV calibration table the tx value for Pixel 2 is -32. The tx power advertised at POWER_LOW is -15. 

I assume the -17 was found in decrypted metadata?

Then it seems to work: 
(-17) + (-15) = -32

Or, if you are a Pixel2 and want to calculate the txPower value for metadata:
- Look up the Pixel2 calibration value --> -32
- Look up your ADVERTISE_TX_POWER_LOW value --> -15
- subtract them: -32 - (-15) = -17

**Comment by horiajurcut** [comment]
@mh- Yes, -17 was found in the decrypted metadata. So the one in the decrypted metadata which you can reverse engineer using the transmission power of the advertisement and the CSV, you can pluck into the attenuation formula. Right?

**Comment by mh-** [comment]
@horiajurcut 
If you now look at the receiver device, where you want to calculate
`Attenuation = TX_POWER - (RSSI_MEASURED + RSSI_CORRECTION)`
you would use
- TX_POWER from decrypted metadata (-17 in the example above)
- RSSI_MEASURED as measured when the RPI beacon was received
- RSSI_CORRECTION from the CSV calibration table, 3rd column "rssi correction"

**Comment by horiajurcut** [comment]
@mh- Yep, that was my guess but it's a bit confusing how these names are being thrown around. Thank you for taking the time to clarify this.

Reverse engineer = I was just glorifying that addition/subtraction operation 😄
